expression/builtin_op.go

```diff
@@ -659,25 +664,113 @@ type isNullFunctionClass struct {
 }
 
 func (c *isNullFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinIsNullSig{newBaseBuiltinFunc(args, ctx)}
-	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
+	if err := c.verifyArgs(args); err != nil {
+		return nil, errors.Trace(err)
+	}
+	tc := args[0].GetType().ToClass()
+	var argTp evalTp
+	switch tc {
+	case types.ClassInt:
+		argTp = tpInt
+	case types.ClassDecimal:
+		argTp = tpDecimal
+	case types.ClassReal:
+		argTp = tpReal
+	default:
+		tp := args[0].GetType().Tp
+		if types.IsTypeTime(tp) {
+			argTp = tpTime
+		} else if tp == mysql.TypeDuration {
+			argTp = tpDuration
+		} else {
+			argTp = tpString
+		}
+	}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpInt, argTp)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	bf.tp.Flen = 1
+	var sig builtinFunc
+	switch argTp {
+	case tpInt:
+		sig = &builtinIntIsNullSig{baseIntBuiltinFunc{bf}}
+	case tpDecimal:
+		sig = &builtinDecimalIsNullSig{baseIntBuiltinFunc{bf}}
+	case tpReal:
+		sig = &builtinRealIsNullSig{baseIntBuiltinFunc{bf}}
+	case tpTime:
+		sig = &builtinTimeIsNullSig{baseIntBuiltinFunc{bf}}
+	case tpDuration:
+		sig = &builtinDurationIsNullSig{baseIntBuiltinFunc{bf}}
+	case tpString:
```

Comment:
I would like to keep this for clearer purpose. In addition, if we add another evalTp in the future, it will fall into the `default` cluster, which may not be what we want.